store/messages.js

```diff
@@ -12,183 +12,233 @@ const SEND_MESSAGE_ERROR = "SEND_MESSAGE_ERROR";
-				original: snapshot.val().translations.original
-			};
-
-			const userLanguage = getState().user.language;
-
-			// if the message was sent by the user it will not be translated
-			if (snapshot.val().senderId !== getState().firebase.auth.uid) {
-				// check if translation to user's language exists
-				if (snapshot.val().translations[userLanguage]) {
-					newMessage.text = snapshot.val().translations[userLanguage];
-					dispatch(addMessage(newMessage));
-				} else {
-					// translate the original message to the language of the user
-					fetch(
-						`https://translation.googleapis.com/language/translate/v2?q=${snapshot.val().message}&target=${getLangKey(
-							userLanguage
-						)}&key=AIzaSyBjkzKxFh39nYubNpXp72NkpG15_FSRWdg`
-					)
-						.then(response => {
-							return response.json();
-						})
-						.then(data => {
-							newMessage.text = data.data.translations[0].translatedText;
-							newMessage.translatedFrom =
-								data.data.translations[0].translatedText !== snapshot.val().message
-									? getLangValue(data.data.translations[0].detectedSourceLanguage)
-									: false;
-
-							dispatch(addMessage(newMessage));
-						});
-				}
-			} else {
-				newMessage.text = snapshot.val().translations.original;
-				dispatch(addMessage(newMessage));
-			}
-		});
-	}
+  // query for all messages for the current chat, and add listener on child_added for new messages
+  if (getState().chats.currentChat) {
+    db.ref(`messages/${getState().chats.currentChat.id}`).on(
+      "child_added",
+      function (snapshot) {
+        // format a message object compatible with GiftedChat, message text not added yet
+        const newMessage = {
+          _id: snapshot.key,
+          user: {
+            _id: snapshot.val().senderId,
+            name: snapshot.val().senderName,
+          },
+          createdAt: snapshot.val().timestamp,
+          original: snapshot.val().translations.original,
+        };
+
+        const userLanguage = getState().user.language;
+
+        // if the message was sent by the user it will not be translated
+        if (snapshot.val().senderId !== getState().firebase.auth.uid) {
+          // check if translation to user's language exists
+          if (snapshot.val().translations[userLanguage]) {
+            newMessage.text = snapshot.val().translations[userLanguage];
+            newMessage.translatedFrom =
+              snapshot.val().translations[userLanguage] !==
+              snapshot.val().message
+                ? snapshot.val().detectedSource
+                : false;
+            dispatch(addMessage(newMessage));
+          } else {
+            // translate the original message to the language of the user
+            fetch(
+              `https://translation.googleapis.com/language/translate/v2?q=${
+                snapshot.val().message
+              }&target=${getLangKey(
+                userLanguage
+              )}&key=AIzaSyBjkzKxFh39nYubNpXp72NkpG15_FSRWdg`
```

Comment:
Dont be exposing the api key to firebase

Reply:
🤦